zpl2.py

```diff
@@ -425,3 +427,94 @@ def graphic_circle(self, right, down, graphic_format):
             stop=self._field_data_stop(),
         )
         self._write_command(command)
+
+    @staticmethod
+    def encode_grf_ascii(pil_image):
+        """Converts a PIL image to a .GRF file encoded in ascii for use
+        with zebra printers.
+        @:return array of strings, each string is a pixel row, containing
+        4 pixel per ascii character
+        """
+
+        def round_up(value, multiple):
+            result = value
+            rest = value % multiple
+            if rest:
+                result += multiple - rest
+            return result
+
+        def get_byte_val(bits, word_len):
+            if len(bits) == 0:
+                return 0
+            else:
+                byte_value = bits[0] << word_len - 1
+                return byte_value + get_byte_val(bits[1:], word_len - 1)
+
+        # make sure we have a monochrome image, and load it.
+        data = pil_image.convert("1").load()
+        sizex, sizey = pil_image.size
+
+        # increase row size to the nearest multiple of 8 pixels
+        # The length of the first row of pixels is set in bytes and must be
+        # a multiple of 8 bytes.
+        # The zebra tools also round the x size of images to the nearest of
+        # 8 pixels.
+        row_size = round_up(sizex, 8)
+        # construct an array y - x
+        # the first row of the array, is the first line of pixels in the
+        # image
+        # Fill all values with white
+        pixels = [[1 for x in range(0, row_size)] for y in range(0, sizey)]
+        for xpos in range(0, sizex):
+            for ypos in range(0, sizey):
+                input_value = data[xpos, ypos]
+                # PIL: pixel value 0 = white , pixel value 255 = black
+                # GRF: 1=white, 0=black
+                bit_value = input_value >> 7 ^ 1
+                pixels[ypos][xpos] = bit_value
+
+        ret = []
+        for pixel_row in pixels:
+            pixel_string = ''
+            for xpos in range(0, len(pixel_row), 4):
+                bits = pixel_row[xpos:xpos + 4]
+                byte_value = get_byte_val(bits, 4)
+                pixel_string += '%X' % byte_value
+            ret.append(pixel_string)
+
+        return ret
+
+    def graphic_image(self, right, down, pil_image):
```

Comment:
IMO this method should be named differently, although I don't have any suggestions without elaboration on what it is doing. As it stands though, the method function is not totally obvious simply by the naming IMO

Reply:
For the API methods naming, I keep the names used in the [official ZPL2 documentation](https://www.zebra.com/content/dam/zebra/manuals/en-us/software/zpl-zbi2-pm-en.pdf) for each ZPL command.

This method should be named `graphic_field` to follow the logic :)

Reply:
Thanks for the elaboration. I agree that alignment to protocol trumps design, any day